Snowflake.Data.Tests/FastParserTest.cs

```diff
@@ -0,0 +1,155 @@
+
+            Assert.Throws<OverflowException>(() => FastParser.FastParseInt32(_byte, 0, _byte.Length));
+        }
+
+        [Test]
+        public void TestFastParseInt32WithNegativeOverflow()
+        {
+            // Int32.MinValue - 1
+            string int32MinValueMinusOne = "-2147483649";
+            _byte = Encoding.UTF8.GetBytes(int32MinValueMinusOne);
+
+            Assert.Throws<OverflowException>(() => FastParser.FastParseInt32(_byte, 0, _byte.Length));
+        }
+
+        [Test]
+        public void TestFastParseInt32ThrowsWrongFormat()
+        {
+            Assert.Throws<FormatException>(() => FastParser.FastParseInt32(new byte[1], 0, 1));
+        }
+
+        [Test]
+        public void TestFastParseDecimalWithLongMaxValuePlusOne()
+        {
+            // Int64.MaxValue + 1
+            ulong int64MaxValuePlusOne = 9223372036854775808;
+            _byte = Encoding.UTF8.GetBytes(int64MaxValuePlusOne.ToString());
+
+            decimal actualDecimalValue = FastParser.FastParseDecimal(_byte, 0, _byte.Length);
+            Assert.AreEqual(int64MaxValuePlusOne, actualDecimalValue);
+        }
+
+        [Test]
+        public void TestFastParseDecimalWithLongMaxValuePlusDecimal()
+        {
+            // Int64.MaxValue + 1.123M
+            decimal int64MaxValuePlusOneWithDecimal = 9223372036854775808.123M;
+            string int64MaxValuePlusOneWithDecimalString = "9223372036854775808.123";
+            _byte = Encoding.UTF8.GetBytes(int64MaxValuePlusOneWithDecimalString);
+
+            decimal actualDecimalValue = FastParser.FastParseDecimal(_byte, 0, _byte.Length);
+            Assert.AreEqual(int64MaxValuePlusOneWithDecimal, actualDecimalValue);
+        }
+
+        [Test]
+        public void TestFastParseDecimalWithPositiveDecimal()
+        {
+            decimal expectedDecimalValue = 1.2345678M;
+            _byte = Encoding.UTF8.GetBytes(expectedDecimalValue.ToString());
+
+            decimal actualDecimalValue = FastParser.FastParseDecimal(_byte, 0, _byte.Length);
+            Assert.AreEqual(expectedDecimalValue, actualDecimalValue);
+        }
+
+        [Test]
+        public void TestFastParseDecimalWithNegativeDecimal()
+        {
+            decimal expectedDecimalValue = -1.2345678M;
+            _byte = Encoding.UTF8.GetBytes(expectedDecimalValue.ToString());
+
+            decimal actualDecimalValue = FastParser.FastParseDecimal(_byte, 0, _byte.Length);
+            Assert.AreEqual(expectedDecimalValue, actualDecimalValue);
+        }
+
+        [Test]
+        public void TestFastParseDecimalWithoutDecimalInTheValue()
+        {
+            decimal expectedDecimalValue = 12345678;
+            _byte = Encoding.UTF8.GetBytes(expectedDecimalValue.ToString());
+
+            decimal actualDecimalValue = FastParser.FastParseDecimal(_byte, 0, _byte.Length);
+            Assert.AreEqual(expectedDecimalValue, actualDecimalValue);
+        }
+    }
```

Comment:
If we pass a null byte array into any FastParser methods an exception System.ArgumentNullException or NullReferenceException is thrown (depending which call has been made).
On the other hand I checked all existing calls to the parser and right now nothing like that should happen (?).
ie.:
`return FastParser.FastParseInt64(srcVal.Buffer, srcVal.offset, srcVal.length);`
which is done after null check for srcVal.
But what if srcVal.Buffer is null whereas srcVal is not null?

My suggestion is to add null checks in FastParser and write just one test showing that parser doesn't break
when one of those parser methods are done with null buffer.

Reply:
I added null checks and return 0 if it's null. Could you take a look?

I'm not sure if it should be returning something else

Reply:
@sfc-gh-igarish @sfc-gh-ext-simba-lf Maybe we should capture that in the parser and instead of NPE or ArgException we should throw SnowflakeDbException with internal error and a message of parsing a null buffer? 
I wonder that returning 0 seams not really a safe solution. We could introduce an error code OR throw exception mentioned above.